### ISSUE TITLE
Steps navigation doesn't work when using Vite bundler

### ISSUE THREAD

**Issue by maltsev** [gilbarbara/react-joyride#769]
## 🐛 Bug Report

When using [Vite bundler](https://vitejs.dev/), navigation between steps and the close step button don't work.
I tested the same code with create-react-app and it works well.

## To Reproduce

See the test repo below.

## Expected behavior

It should work in Vite.

## Link to repl or repo (highly encouraged)

I prepared a test repo: https://github.com/maltsev/react-joyride-debug

Here is also the screencasts (create-react-app and Vite):

https://user-images.githubusercontent.com/923973/147825863-28fca654-200a-46d6-b1e5-61f537d2a43d.mp4

https://user-images.githubusercontent.com/923973/147825871-bc8ba306-e884-400b-8cb2-fb2ed72f630d.mp4

## Run `npx envinfo --system --binaries --npmPackages react-floater`

Paste the results here:

```bash
  System:
    OS: macOS 12.0.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 459.87 MB / 16.00 GB
    Shell: 3.3.1 - /usr/local/bin/fish
  Binaries:
    Node: 17.3.0 - /usr/local/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 8.3.0 - /usr/local/bin/npm
    Watchman: 2021.11.08.00 - /usr/local/bin/watchman
```

**Comment by gilbarbara** [comment]
hey @maltsev 

There are some users using this library with vite without problems.
Can you create a minimal reproduction app to make sure isn't something with your configuration?

**Comment by maltsev** [comment]
Thanks for your quick response!

> Can you create a minimal reproduction app to make sure isn't something with your configuration?

Sure, I tried to do it here: https://github.com/maltsev/react-joyride-debug

It's an example for your docs that I modified a bit to show the tooltips on page load. I used the default CRA and Vite config.

Also, wish you a happy New Year :-)

**Comment by gilbarbara** [comment]
hey,

Thanks, happy new year!

My findings so far:

- removing the `React.StrictMode` fix it.
- Also works without the `stepIndex` even with `StrictMode` (do you really need to set the index manually?)
- It appears that this [conditional](https://github.com/gilbarbara/react-joyride/blob/main/src/components/index.js#L164) is getting a different result with vite compared to CRA.

I don't know what vite is doing in the background but I'm thinking it is a problem with it, or maybe esbuild?

**Comment by maltsev** [comment]
> removing the `React.StrictMode` fix it.

I tried it, but it didn't help. Did you use something like 

```diff
diff --git a/src/index.jsx b/src/index.jsx
index 3b93f09..7d56277 100644
--- a/src/index.jsx
+++ b/src/index.jsx
@@ -4,11 +4,11 @@ import './index.css';
 import App from './App';
 
 ReactDOM.render(
-  <React.StrictMode>
+  <div>
     <App />
     <button className="btn my-first-step">first</button>
     <button className="btn my-second-step">second</button>
     <button className="btn my-third-step">third</button>
-  </React.StrictMode>,
+  </div>,
   document.getElementById('root')
 );

```

?


> do you really need to set the index manually?

Unfortunately, yes. I want to display the first unseen tooltip on page load but also let the user navigate to the previous  seen tooltips.


> I don't know what vite is doing in the background but I'm thinking it is a problem with it, or maybe esbuild?

Yes, it's definitely something with Vite or its internal tools. I'm still not sure in what capacity Vite uses esbuild as their documentation states that they use [rollup for bundling](https://vitejs.dev/guide/why.html#why-not-bundle-with-esbuild).

---

Thanks for looking into that! I'll keep you updated about my attempts to solve it!

**Comment by ccelent** [comment]
I'm running into a near identical issue when migrating to webpack.  My previously working controlled-mode component no longer works after the swap in the same way noted above.  You can click skip, but close and next both are completely unresponsive and never invoke the callback.

Moving to non-controlled mode does work better, but has some possibly-unrelated other issues, such as the inability to replay the tutorial. 

I just wanted to chime in since I was working on troubleshooting a similar problem and wanted to note it doesn't seem unique to vite.

**Comment by valtism** [comment]
I also have this problem moving off Webpack to Vite. I am not, however, using strict mode.

**Comment by valtism** [comment]
The issue seemed to be the way that the `tree-changes` package was importing from the `fast-deep-equal` library. `fast-deep-equal` is not esm friendly, so since Vite has issues with non-esm, it would mean that calling `equal` would throw. 

Unfortunately the underlying error would be caught meaning surfacing the real error took some investigation. Still, I think the fix linked above should solve this issue.

This was the last (I hope) blocker to migrating our app from CRA to Vite, so will be keen to see this fixed after @gilbarbara has had a chance to look it over and make sure it all works.

**Comment by valtism** [comment]
So the fix I proposed doesn't look like it is going to work for the library, as outlined here: https://github.com/gilbarbara/tree-changes/pull/18#issuecomment-1009877162

Since the project I am using this library with already has `esModuleInterop` enabled, I will be using the linked fix by using the `patch-package` library. If you want to do this as a workaround you can install `patch-package` and add this patch inside a top-level `patches` folder to make things work in Vite: https://gist.github.com/valtism/f423ed3ae1f2ea886d221b575970daec

**Comment by erm1116** [comment]
I'm facing this issue and try the workaround using `patch-package` and works it.
But, I hope this problem will be fixed.

@gilbarbara 
I saw your `tree-changes` package has been supported esm builds.
https://github.com/gilbarbara/tree-changes/pull/20

Do you have any plans to fix this issue?
(maybe, I think this issue is fixed to upgrade `tree-changes` package version to 0.9.0)

**Comment by pierre-borckmans** [comment]
@erm1116 @gilbarbara For me, updating `tree-changes` to `0.9.0` indeed fixes this issue. 
I opened a PR for it here: #786 
Let me know if yo think this change is suitable.
Thanks!

**Comment by gilbarbara** [comment]
Fixed in 2.4.0
